workspaces/ui-v2/src/types/shapes.ts

```diff
@@ -0,0 +1,62 @@
+import { ChangeType } from './changes';
+import { IContribution } from './contributions';
+
+// Types for rendering shapes and fields
+export interface IFieldRenderer {
+  fieldId: string;
+  name: string;
+  shapeId: string;
+  shapeChoices: IShapeRenderer[];
+  required: boolean;
+  changes: ChangeType | null;
+  contributions: Record<string, string>;
+  additionalAttributes?: string[];
+}
+
+// Used to render an objects field details and contributions
+export interface IFieldDetails {
+  name: string;
+  contribution: IContribution;
+  shapes: IShapeRenderer[];
+  depth: number;
+}
+
+// Used to render query parameters
+export type QueryParameters = Record<string, IFieldRenderer>;
+
+export type IShapeRenderer =
```

Comment:
What's important to note here is there are three types of shapes (previously we had one, which meant if we added properties after we didn't have great type safety since we had to make things `optional`).
- SpectacleShape (shape queried from spectacle)
- ReduxShape (shape that's stored in redux)
- IShapeRenderer (shape that's used to render in components)